exonum/src/node/mod.rs

```diff
@@ -809,34 +809,25 @@ pub struct ConnectInfo {
     pub public_key: PublicKey,
 }
 
-impl ConnectInfo {
+impl StorageValue for ConnectInfo {
     /// Tries to serialize the given `ConnectInfo` into vector of bytes.
-    pub fn try_serialize(self) -> Vec<u8> {
+    fn into_bytes(self) -> Vec<u8> {
         let mut vec_bytes = vec![0_u8; Self::field_size() as usize];
         self.write(&mut vec_bytes, 0, Self::field_size());
         vec_bytes
     }
 
     /// Tries to deserialize the given `value` into `ConnectInfo`.
     #[allow(unsafe_code)]
-    pub fn try_deserialize(value: &[u8]) -> Self {
-        unsafe { Self::read(&value, 0, Self::field_size()) }
-    }
-}
-
-impl StorageValue for ConnectInfo {
-    fn into_bytes(self) -> Vec<u8> {
-        self.try_serialize()
-    }
-
     fn from_bytes(value: Cow<[u8]>) -> Self {
-        Self::try_deserialize(value.as_ref())
+        assert_eq!(value.len(), Self::field_size() as usize);
+        unsafe { Self::read(value.as_ref(), 0, Self::field_size()) }
```

Comment:
I believe there are more things to validate here, aside from size. One of `ConnectInfo`'s fields is `SocketAddr` which is likely to have complex serialized structure that can easily be corrupted. It is used to hold IPv4 or IPv6 addresses, but what if the bytes are pure garbage and denote neither? Yes, the process is unlikely to segfault due to size assertion, but I guess networking code can start behaving in weird ways (or fail unexpectedly) if it encounters unintelligible `struct sockaddr`.

I guess we can roll with `unsafe` here if we give some pinky promise to never feed invalid data into this function, but I strongly prefer more explicit runtime checks over random crashes or undefined behavior resulting from `unsafe` blocks not being actually safe. Intended usage of `StorageValue` trait (serialize/deserialize values for blockchain storage) can act as such pinky promise, I guess, as you're supposed to use `from_bytes` only on slices resulting from `into_bytes`. And there is a precedent of `Duration` using `unsafe` for `StorageValue` implementation.

Thus I do not consider this serious enough to prevent the PR from being merged. If there is no low-effort way to avoid this unsafe block then go ahead. That's just my personal dislike of unmotivated `unsafe` blocks in cases where they should have been avoidable. Probably we should have a task to audit and clean up `unsafe` blocks at some point.

Reply:
> Probably we should have a task to audit and clean up unsafe blocks at some point.

And I should definitely learn to read before I write :man_facepalming: as we do have such tasks:

* [[ECR-174] Revise StorageKey interface](https://jira.bf.local/browse/ECR-174)
* [[ECR-641] Safety of StorageValue/StorageKey read methods](https://jira.bf.local/browse/ECR-641)
* [[ECR-800] Revise unsafe code](https://jira.bf.local/browse/ECR-800)

some of which even explicitly refer to the these exact issues of `StorageValue` trait.

So please go ahead with the merge, this particular `unsafe` block may be fixed later.